pkg/querier/batch/batch.go

```diff
@@ -85,49 +86,79 @@ func newIteratorAdapter(underlying iterator) chunkenc.Iterator {
 }
 
 // Seek implements chunkenc.Iterator.
-func (a *iteratorAdapter) Seek(t int64) bool {
+func (a *iteratorAdapter) Seek(t int64) chunkenc.ValueType {
 
 	// Optimisation: fulfill the seek using current batch if possible.
 	if a.curr.Length > 0 && a.curr.Index < a.curr.Length {
 		if t <= a.curr.Timestamps[a.curr.Index] {
 			//In this case, the interface's requirement is met, so state of this
 			//iterator does not need any change.
-			return true
+			return a.curr.ValueTypes
 		} else if t <= a.curr.Timestamps[a.curr.Length-1] {
 			//In this case, some timestamp between current sample and end of batch can fulfill
 			//the seek. Let's find it.
 			for a.curr.Index < a.curr.Length && t > a.curr.Timestamps[a.curr.Index] {
 				a.curr.Index++
 			}
-			return true
+			return a.curr.ValueTypes
 		}
 	}
 
 	a.curr.Length = -1
 	a.batchSize = 1
-	if a.underlying.Seek(t, a.batchSize) {
+	if typ := a.underlying.Seek(t, a.batchSize); typ != chunkenc.ValNone {
 		a.curr = a.underlying.Batch()
-		return a.curr.Index < a.curr.Length
+		if a.curr.Index < a.curr.Length {
+			return typ
+		}
 	}
-	return false
+	return chunkenc.ValNone
 }
 
 // Next implements chunkenc.Iterator.
-func (a *iteratorAdapter) Next() bool {
+func (a *iteratorAdapter) Next() chunkenc.ValueType {
 	a.curr.Index++
-	for a.curr.Index >= a.curr.Length && a.underlying.Next(a.batchSize) {
+	for a.curr.Index >= a.curr.Length && a.underlying.Next(a.batchSize) != chunkenc.ValNone {
 		a.curr = a.underlying.Batch()
 		a.batchSize = a.batchSize * 2
 		if a.batchSize > chunk.BatchSize {
 			a.batchSize = chunk.BatchSize
 		}
 	}
-	return a.curr.Index < a.curr.Length
+
+	if a.curr.Index < a.curr.Length {
+		return a.curr.ValueTypes
+	}
+	return chunkenc.ValNone
 }
 
 // At implements chunkenc.Iterator.
 func (a *iteratorAdapter) At() (int64, float64) {
-	return a.curr.Timestamps[a.curr.Index], a.curr.Values[a.curr.Index]
+	return a.curr.Timestamps[a.curr.Index], a.curr.SampleValues[a.curr.Index]
+}
+
+// AtHistogram implements chunkenc.Iterator.
+func (a *iteratorAdapter) AtHistogram() (int64, *histogram.Histogram) {
+	return a.curr.Timestamps[a.curr.Index], a.curr.HistogramValues[a.curr.Index]
+}
+
+// AtFloatHistogram implements chunkenc.Iterator.
+func (a *iteratorAdapter) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
+	// It seems this method is sometimes called when there aren't any float histograms
+	// but there are regular histograms, so we can get of those instead.
```

Comment:
Is this your comment or you brought it from somewhere? Maybe it's worth adding a `TODO revisit for native histogram`?

Reply:
This comes from sparsehistogram branch. The reason I think is that promql evaulation prefers to return FloatHistogram always, as that is easier to combine and work with later, e.g.: https://github.com/prometheus/prometheus/blob/fbbc04433416c4114942d8f319fe906d873ab18c/promql/engine.go#L1771

@beorn7 @codesome to confirm and I update the comment

Reply:
> promql evaulation prefers to return FloatHistogram always

This is right. PromQL always works on the FloatHistogram and results in a FloatHistogram